cirq-core/cirq/sim/density_matrix_simulator.py

```diff
@@ -320,17 +313,7 @@ def set_density_matrix(self, density_matrix_repr: Union[int, np.ndarray]):
             mixed state it must be correctly sized and positive semidefinite
             with trace one.
         """
-        if self._split_untangled_states:
-            # TODO: Fix in #4110
-            raise ValueError(  # coverage: ignore
-                'Cannot set states when using `split_untangled_states` option.'  # coverage: ignore
-            )  # coverage: ignore
-        density_matrix = qis.to_valid_density_matrix(
-            density_matrix_repr, len(self._qubit_map), qid_shape=self._qid_shape, dtype=self._dtype
-        )
-        sim_state_matrix = self._simulator_state().density_matrix
-        density_matrix = np.reshape(density_matrix, sim_state_matrix.shape)
-        np.copyto(dst=sim_state_matrix, src=density_matrix)
+        self._sim_state = self._simulator._create_act_on_args(density_matrix_repr, self._qubits)
```

Comment:
I'm not 100% sold on this being the best way to implement this. Passing in the simulator itself here seems heavy. I want to think about alternatives first.

Reply:
No, I think this is okay. This allows creating optimal act_on_args type if it's a pure state. And I think it's actually best to explicitly update sim_state in core_iterator, if we are going to support the functionality at all. Having it update implicitly is more fragile.